compiler/rustc_middle/src/mir/mod.rs

```diff
@@ -2267,6 +2267,9 @@ pub enum Rvalue<'tcx> {
     /// away after type-checking and before lowering.
     Aggregate(Box<AggregateKind<'tcx>>, Vec<Operand<'tcx>>),
 
+    /// Creates a value of the variant with all fields uninitialized
+    UninitVariant(Ty<'tcx>, VariantIdx),
```

Comment:
What does this even *mean* for niche-optimized enums?

`write_discriminant` is a NOP for the 'dataful' variant of a niche-optimized enum, so e.g. `UninitVariant` on the `Some` variant of an `Option<&T>` would simply create a fully uninit value (as implemented for CTFE/Miri by this PR). Is that the intention?

Reply:
Yes, that is *exactly* the intention. My hope is to soon start emitting these things as a part of struct deaggregation as well, where we would basically be in the same boat; an `UninitVariant` for this is just a fully uninit value.